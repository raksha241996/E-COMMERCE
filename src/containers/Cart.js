import React, { Component } from 'react';
import '../styles/containerStyles/Cart.scss';


export default class Cart extends Component {
    constructor(props){
        super(props);
        this.state = {
            items : []
            };

    }

    componentDidMount(){
        this.setState({items : this.getLocal(), count : this.getCount()});   
    }

    componentWillUnmount() {
        this.setLocal(this.state.items);
    }

    getLocal = () => JSON.parse(localStorage.getItem('cart')) 
    setLocal = (items) => {localStorage.setItem('cart',JSON.stringify(items));}
    getCount = () => parseInt(localStorage.getItem('count'))

    increaseByOne = (id) => {
        let items = this.state.items;
        let ite = items.findIndex(x =>  x.id === id);
        items[ite].quantity= parseInt(items[ite].quantity)+1;
        this.setState({items : items, count : this.getCount()});
        this.setLocal(this.state.items);
    }
    
    
    decreaseByOne = (id) => {
        let items = this.state.items;
        let ite = items.findIndex(x =>  x.id === id);
        
        if(items[ite].quantity > 1){
            items[ite].quantity= parseInt(items[ite].quantity)-1;  
        }
        else {
            alert("Click Remove to delete the item")
        }     
        this.setState({items : items, count : items.length});
        this.setLocal(this.state.items);
    }

    removeItem = (id) => {
        let items = this.state.items;
        items = items.filter(x => x.id !== id);
        this.setLocal(items);
        localStorage.setItem('count', items.length); 
        this.props.handleCount(items.length);
        this.setState({items : this.getLocal()});
        
    }

    CalculatePrice = (props) => {
        let total = 0;
        let items = props.item

        for(let i=0;i<items.length;i++){
            total += (parseInt(items[i].Price)*parseInt(items[i].quantity))
        }

        return (
            <article id="calculations">
                <p id="countItem">Total {items.length} items</p>
                <section id="line"><hr /></section>
                <p id="totalPrice">Total Amount : ${total}</p>
            </article>
        )
    }

    CreateListoOfItems = (props) => {
        let itemList = [];            
        
        props.item.forEach((item) => {
            itemList.push(
                    <article id="item" key = {item.id}>
                        <article id="imgndisnbtn">
                            <img src={item.img} alt="item"/>
                                <section id="item-section">
                                    <p id="discr">{item.description}</p>
                                    <p id="quant">Quantity : {item.quantity}</p>
                                    <p id="price">Price : $  {item.Price}</p>    
                                </section>                      
                            <button onClick ={() => {this.removeItem(item.id)}}>Remove</button>
                            
                        </article>
                        <article id="btns">
                            <button id="plus" onClick = {() => {this.increaseByOne(item.id)}}>+</button>
                            <button id="minus" onClick ={() => {this.decreaseByOne(item.id)}}>-</button>
                        </article>    
                    </article>    
            )
        });

        return itemList;
    }

  

    
    render() {
        return (
            <section className="CartBody">                
                    <section id="itemListDisplay">
                        <section id="flexContainer">
                            <this.CreateListoOfItems item={this.state.items}/>                                                                                         
                        </section>
                    </section>
                    {this.state.items.length ?
                        <aside id="priceDetails">
                            <article id="priceHead">Price Details </article>
                            <this.CalculatePrice item={this.state.items} />                        
                        </aside> 
                           :
                        <p id="noItems">No Items In Cart !!</p> }
                </section>
        )
    }
}

