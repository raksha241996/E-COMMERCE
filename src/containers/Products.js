import React from 'react';
import DisplayProduct from '../components/DisplayProduct';
import Pagination from '../components/Pagination';
import { connect } from 'react-redux'
import productsAction from '../action/productsAction'
import { bindActionCreators } from '../../../../Library/Caches/typescript/3.5/node_modules/redux';

export class Products extends React.Component {


    constructor(props) {
        super(props)

        this.state = {
            disable: false,
            productsArray: [],
            currentPage: 1,
            ProductsPerPage: 3,
            currentProducts: 0
        }

        this.cart = []
        const indexOfLastProduct = 3;
        const indexOfFirstProduct = 2;


    }

    componentDidMount() {
        this.getDataCall();
        console.log("inside products page " + this.props.products)
        this.cart = JSON.parse(localStorage.getItem('cart'));
        
    }

    getDataCall = () => {
        //making API call here
        this.props.fetchProducts().then(() => {

            console.log("getData" + this.props.products);

            this.setState({

                productsArray: this.props.products,
            })
            this.updateContent();
        }
        );

    }

    updateContent = () => {
        this.indexOfLastProduct = this.state.currentPage * this.state.ProductsPerPage;
        this.indexOfFirstProduct = this.indexOfLastProduct - this.state.ProductsPerPage;
    }


    paginate(num,e) {

       
        this.setState({ currentPage: num }, () => {
            this.updateContent();
        });

    }


    updateQuantity(product) {
        let index = this.cart.findIndex(x => x.id === product.id);
        console.log(product.id, "index is ", index)
        if (index === -1) {
            this.cart.push(product);
        }
        else {
            this.cart[index].quantity = parseInt(this.cart[index].quantity) + 1;
        }
        localStorage.setItem('count', this.cart.length);
        this.props.handleCount(this.cart.length);

    }

    toggleButton(products) {
        this.setState({ disable: true })
        alert("Item Added");
        this.updateQuantity(products);
        localStorage.setItem("cart", JSON.stringify(this.cart));

    }


    render (){

        return (
            <section>
                <Pagination
                    postsPerPage={this.state.ProductsPerPage}
                    totalPosts={this.props.products.length}
                    paginate={this.paginate.bind(this)} />
                <DisplayProduct products={this.props.products.slice(this.indexOfFirstProduct, this.indexOfLastProduct)} toggleButton={this.toggleButton.bind(this)} />

            </section>

        );


    }
}


// mapDispatchToProps is used for dispatching actions to the store 
const mapDispatchToProps = (dispatch) => {
    return bindActionCreators(productsAction, dispatch)
}


//to get the value from store and assign it to props
const mapStateToProps = state => {
    return {
        products: state.products.products
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Products)

