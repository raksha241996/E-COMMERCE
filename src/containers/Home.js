import React from 'react';
import welcomeimage from '../assets/welcomeimage.jpg'
import '../styles/containerStyles/HomeStyle.scss'
import Button from '@material-ui/core/Button';
import { BrowserRouter as Router, Route, Link } from "react-router-dom";
import Products from "../containers/Products"
export default function Home() {

    

    return (
      
     
    <div className="Home">
        <img src={welcomeimage} alt="welcome" width="100%" >
        </img>
        {/* <button className="btn">Start Shopping</button> */}
        {/* <Button variant="contained" color="primary" className="btn" >
        Start Shopping !!
      </Button> */}
      <Route path="/products/" component={Products} />
    </div>
    
    
        );
  }