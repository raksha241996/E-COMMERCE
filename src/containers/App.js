import React from 'react';
import '../styles/containerStyles/App.scss'
import { BrowserRouter as Router, Route, Link } from "react-router-dom";

import Home from './Home'
import Products from './Products'
import HomeIcon from '@material-ui/icons/Home'
import ShopIcon from '@material-ui/icons/Shop'
import ShoppingCartIcon from '@material-ui/icons/ShoppingCart'
import Cart from './Cart';

class App extends React.Component {
  constructor(props){
    super(props);

    this.state ={
      count : 0
    };

    this.handleCount = this.handleCount.bind(this);
  }

  componentWillUpdate(){
    parseInt(localStorage.getItem('count')) === 0 ? this.hideCount() : this.showCount();
  }

  componentDidMount(){
    this.setState({count : parseInt(localStorage.getItem('count')) })
  }

  handleCount(count){
    this.setState({count : count}); 
  }

  hideCount(){
    document.querySelector("#count").style.visibility = 'hidden';
  }
  showCount(){
    document.querySelector("#count").style.visibility = 'visible';
  }

  render() {
    return (
      
      <Router>
        <header>
          <nav className="topnav">
            <a href="#home" tabIndex="1" ><Link to="/"><HomeIcon/>Home</Link></a>
            <a href="#produts" tabIndex="1"> <Link to="/products/"><ShopIcon/>Products</Link></a>
            <a href="#cart"><Link to="/cart/"><ShoppingCartIcon/>Cart</Link> <span id="count">{ this.state.count }</span> </a>
          </nav>
        </header>
        <Route path="/" exact component={Home} />
        <Route path="/products/"  render={() =>  <Products handleCount = {this.handleCount} /> } />
        <Route path="/cart/" render={() =>  <Cart handleCount = {this.handleCount} />  } />
      </Router>
    
    );
  }
}

export default App;
