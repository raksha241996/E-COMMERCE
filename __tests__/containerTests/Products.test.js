import React from 'react';
import { shallow ,configure} from 'enzyme';
import {Products} from '../../src/containers/Products'
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

//mocking the store for testing
const store = mockStore({
 dispatch: jest.fn(),
        products:{
     }
    
});



//configuring the Adapter for using enzyme
configure({ adapter: new Adapter() });

//creating static props for testing
const createTestProps = props => ({
    products :[{}],
    handleCount:jest.fn(),
    fetchProducts:jest.fn().mockImplementation(()=>{
        return Promise.resolve({});
    })
})


describe('Product component test with Enzyme', () => {
    it('renders without crashing', () => {
        const props = createTestProps({});
       const wrapper =   shallow(  
           <Products store={store} {...props}/>
         );
          const instance = wrapper.instance()
          expect(wrapper).toMatchSnapshot();
     });
 });


 describe('Product component test with Enzyme', () => {
    it('toggleButton calls without crashing', () => {
        const props = createTestProps({});
       const wrapper =   shallow(  
           <Products store={store} {...props}/>
         );
          const instance = wrapper.instance()
          instance.cart=[]
          instance.toggleButton(props.products)
          expect(wrapper).toMatchSnapshot();
     });
 });