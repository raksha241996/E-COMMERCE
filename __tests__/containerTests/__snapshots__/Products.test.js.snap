// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component test with Enzyme renders without crashing 1`] = `ShallowWrapper {}`;

exports[`Product component test with Enzyme renders without crashing 1`] = `ShallowWrapper {}`;

exports[`Product component test with Enzyme toggleButton calls without crashing 1`] = `ShallowWrapper {}`;
